src/context/AppContextProvider.tsx

```diff
@@ -0,0 +1,256 @@
+
+      setTodos(updatedTodos as Todo[]);
+    } catch (error) {
+      setErrorMessage(ErrorMessage.UPDATE);
+      setShowError(true);
+      throw new Error(ErrorMessage.UPDATE);
+    } finally {
+      setTodosBeingLoaded(prev => prev.filter(id => id !== todo.id));
+    }
+  };
+
+  // TOGGLE TODO STATUS
+  const changeTodoStatus = async (todoId: number, todoStatus: boolean) => {
+    setTodosBeingLoaded(prev => ([...prev, todoId]));
+
+    try {
+      const response = await patchTodo(todoId, { completed: !todoStatus });
+
+      setTodos(prevTodos => (
+        prevTodos.map(item => (
+          item.id === todoId
+            ? response
+            : item
+        )) as Todo[]
+      ));
+    } catch (error) {
+      setErrorMessage(ErrorMessage.UPDATE);
+      setShowError(true);
+    } finally {
+      setTodosBeingLoaded(prev => prev.filter(id => id !== todoId));
+    }
+  };
+
+  // CHANGING FILTER
+  const changeFilter = (event: MouseEvent<HTMLAnchorElement>) => {
+    const { id } = event.target as HTMLAnchorElement;
+
+    if (selectedFilter === id) {
+      return;
+    }
+
+    setSelectedFilter(id as Filter);
+  };
+
+  // CLEAR ALL COMPLETED
+  const clearCompleted = async () => {
+    const completedTodos: Todo[] = todos.filter(todo => todo.completed);
+
+    setTodosBeingLoaded(prev => ([
+      ...prev,
+      ...completedTodos.map(todo => todo.id),
+    ]));
+
+    completedTodos.map(async (todo) => {
+      try {
+        await deleteTodo(todo.id);
+        setTodos(prev => prev
+          .filter(todoToDelete => todoToDelete.id !== todo.id));
+      } catch (error) {
+        setErrorMessage(ErrorMessage.DELETE);
+        setShowError(true);
+      } finally {
+        setTodosBeingLoaded([]);
+      }
+    });
+  };
+
+  // TOGGLE ALL
+  const toggleAllStatus = async () => {
+    try {
+      let updatedTodos;
+
+      if (todos.length === completedTodosNum) {
+        setTodosBeingLoaded(todos.map(todo => todo.id));
+        const responses = await Promise.all(
+          todos.map(todo => patchTodo(todo.id, { ...todo, completed: false })),
+        );
+
+        updatedTodos = responses as Todo[];
+      } else {
+        const todosToUpdate = todos.filter(todo => !todo.completed);
+
+        setTodosBeingLoaded(todosToUpdate.map(todo => todo.id));
+        await Promise.all(
+          todosToUpdate.map(todo => patchTodo(
+            todo.id, { ...todo, completed: true },
+          )),
+        );
+
+        updatedTodos = todos.map(todo => (
+          todo.completed
+            ? todo
+            : { ...todo, completed: true }));
+      }
+
+      setTodos(updatedTodos);
+    } catch (error) {
+      setErrorMessage(ErrorMessage.UPDATE);
+      setShowError(true);
+    } finally {
+      setTodosBeingLoaded([]);
+    }
+  };
+
+  // EFFECTS
+  useEffect(() => {
+    loadData();
+  }, []);
+
+  useEffect(() => {
+    if (showError) {
+      setTimeout(() => {
+        setShowError(false);
+      }, 3000);
```

Comment:
will be better if Error component hold this functionality (hide after 3sec)

Reply:
done ✅